DIPs/DIP1038.md

```diff
@@ -10,11 +10,11 @@
 
 ## Abstract
 
-This DIP proposes a new attribute, `@nodiscard`, that allows the programmer to
-make ignoring a function's return value into a compile-time error. It can be
-used to implement alternative error-handling mechanisms for code that cannot
-use exceptions, and to prevent bugs when interfacing with external functions
-that report errors via their return values.
+This DIP proposes a new attribute, `@mustUse`, which can be applied to a
+`struct`, `union`, or `enum` type to make ignoring an expression of that type
```

Comment:
Does this DIP imply an restrictions on the base type of the enum, i.e. is something like this allowed?

```d
@mustUse
enum Important : Object;
```

Reply:
Currently it's allowed, but I guess it probably shouldn't be. Good catch.

Reply:
Actually, if we follow the same logic as for classes, a `@mustUse` enum should only be allowed if its base type is also `@mustUse`. Which would make putting`@mustUse` on the enum itself kind of pointless. So maybe the answer is to remove enuns from the DIP entirely.

Reply:
The comparison seems lackluster given that an `enum` simply denotes a *fixed* subset of values of the base type - but I'll have to think more about that.

Removing `enum` support would hurt in practice given that return codes are a primary use-case of `@mustUse`.

Reply:
All that matters for the comparison is that classes and enums both implicitly convert to their base type. Allowing `@mustUse` on enums would make it possible to write code like the following:

```d
@mustUse enum E : int { x }
E f() { return E.x; }
int g() { return f(); }
void ignore() { g(); } // oops - implicitly discards E.x
```

This is exactly the failure mode Walter said was unacceptable for classes.

Reply:
> All that matters for the comparison is that classes and enums both implicitly convert to their base type. Allowing 

Okay but then `alias this` must be forbidden for `@mustUse struct`'s to be consistent.

`@mustUse` on enums would make it possible to write code like the following:

Guess that boils down to the rather contrieved definition of "using a `@mustUse` value".  The return value of `f` was used according to the definition in this DIP, `g` simply decided to weaken the guarantees of the returned value. 

But requiring an explicit cast / unwrap might be beneficial - even though the end result would be identical in that example.

Reply:
The fact that the programmer has to explicitly write an `alias this` declaration makes it a little less clear-cut. I'm inclined to allow it, simply for the sake of avoiding special cases, but there are arguments to be made on both sides.